### PR TITLE
Jdk19 build with Jetty 12

### PR DESCRIPTION
**Pull request by olamy**
Signed-off-by: Olivier Lamy <olamy@apache.org>

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -44,6 +44,40 @@ pipeline {
...
         }
+        stage("Build / Test - JDK19") {
+          agent { node { label 'linux' } }
+          steps {
+            container('jetty-build') {
+              timeout( time: 180, unit: 'MINUTES' ) {
+                checkout scm
+                mavenBuild( "jdk19", "clean install javadoc:javadoc -Perrorprone", "maven3")
+                // Collect up the jacoco execution results (only on main build)
+                jacoco inclusionPattern: '**/org/eclipse/jetty/**/*.class',
```
Comment (joakime) [inline Jenkinsfile:55]: I think we can skip jacoco for this build.

Comment (joakime) [reply Jenkinsfile:55]: Not sure it even works on jdk19 anyway.

Comment (olamy) [reply Jenkinsfile:55]: certainly not for now but not the main concern of this PR :)

**Comment by olamy** [comment]
@joakime @lorban @sbordet 

as expected with jdk 19 
```

[ERROR] Error occurred during initialization of boot layer

[ERROR] java.lang.module.FindException: Module jdk.incubator.foreign not found, required by org.eclipse.jetty.quic.quiche.foreign.incubator
```

2 solutions:
1. get rid of the maven module `jetty-quic-quiche-foreign-incubator` which doesn't work anymore in 19 (as the package is now `java.lang.foreign` and simply write a `jetty-quic-quiche-foreign`  (obviously only available in a profile `[19,)`
2. keep the maven module `jetty-quic-quiche-foreign-incubator` but fix the activation from `<jdk>[17,)</jdk>` to `<jdk>17</jdk>` and write another new module `jetty-quic-quiche-foreign`  (obviously only available in a profile `[19,)`  

I'm voting for 1 (I don't see the need of keeping this jdk17 incubator compatibility)

build should pass now on jdk 19 with this commit https://github.com/eclipse/jetty.project/pull/8707/commits/146c100372e309d30b260245754893963874aaab

but still what should we do?

**Comment by sbordet** [comment]
@olamy we want to keep the module for all those that run Jetty with Java 17 because it's an LTS.

So please don't do 1, but do 2 instead.
I thought @lorban fixed this days ago?

**Comment by olamy** [comment]
> @olamy we want to keep the module for all those that run Jetty with Java 17 because it's an LTS.
> 
> So please don't do 1, but do 2 instead. I thought @lorban fixed this days ago?

**Comment by joakime** [comment]
The foreign incubator issue was fixed in PR #8699
@lorban made it so that the quiche-client only uses the foreign incubator on jdk 17 (exclusive)

**Comment by olamy** [comment]
@joakime but we still need to fix/add the build for jdk19 which is the goal of this pr.
why closing it?

**Comment by joakime** [comment]
I thought I reopened it.
Shoot.  Opening it again.

**Comment by olamy** [comment]
weird it looks I closed it first.  
LOL lesson learned never look at PR after G&T 😜

**Comment by lorban** [comment]
Are you seeing the build fail with JDK 19? If so, we're facing a "it works on my machine" problem and we should discuss this in details.

**Comment by olamy** [comment]
without this change https://github.com/eclipse/jetty.project/commit/146c100372e309d30b260245754893963874aaab the build definitely fail with jdk19.
`<jdk>[17,)</jdk>` means the profile is activated with 19 so it include the dependency 

```
        <dependency>
          <groupId>org.eclipse.jetty.quic</groupId>
          <artifactId>jetty-quic-quiche-foreign-incubator</artifactId>
        </dependency>
```

which is using the package `import jdk.incubator.foreign` and try to import the module `requires jdk.incubator.foreign;` which doesn;t exist anymore with 19 so the failure here https://jenkins.webtide.net/blue/organizations/jenkins/jetty.project/detail/PR-8707/1/pipeline

```

[ERROR] Error occurred during initialization of boot layer

[ERROR] java.lang.module.FindException: Module jdk.incubator.foreign not found, required by org.eclipse.jetty.quic.quiche.foreign.incubator
```

**Comment by lorban** [comment]
Problem: with your change, the [jetty-quic-quiche-foreign-incubator](https://github.com/eclipse/jetty.project/tree/jetty-12.0.x/jetty-core/jetty-quic/jetty-quic-quiche/jetty-quic-quiche-foreign-incubator) module doesn't get built when the build runs on JDK 19. This is a mistake: this module should be built in all cases, and JDK 19 is capable of compiling the code so it can generate the artifact.

But JDK 18 and above cannot run the code, so the tests should only be executed on JDK 17. This is why [the module's pom](https://github.com/eclipse/jetty.project/blob/jetty-12.0.x/jetty-core/jetty-quic/jetty-quic-quiche/jetty-quic-quiche-foreign-incubator/pom.xml) contains a JDK 18+ profile that is meant to disable surefire tests. Apparently this doesn't work as well as it should.

**Comment by olamy** [comment]
> Problem: with your change, the [jetty-quic-quiche-foreign-incubator](https://github.com/eclipse/jetty.project/tree/jetty-12.0.x/jetty-core/jetty-quic/jetty-quic-quiche/jetty-quic-quiche-foreign-incubator) module doesn't get built when the build runs on JDK 19. This is a mistake: this module should be built in all cases, and JDK 19 is capable of compiling the code so it can generate the artifact.
> 

the module `jetty-quic-quiche-foreign-incubator` doesn't get build because of this https://github.com/eclipse/jetty.project/blob/76ca392917fa58811b4f2b72eb6b8a988a39fdf0/jetty-core/jetty-quic/jetty-quic-quiche/pom.xml#L23 (nothing related to my change ;) )

my change just avoid the test to fail on jdk19 by introducing a dependency on it.

I will change this value in this PR to `[17,)` as well

> But JDK 18 and above cannot run the code, so the tests should only be executed on JDK 17. This is why [the module's pom](https://github.com/eclipse/jetty.project/blob/jetty-12.0.x/jetty-core/jetty-quic/jetty-quic-quiche/jetty-quic-quiche-foreign-incubator/pom.xml) contains a JDK 18+ profile that is meant to disable surefire tests. Apparently this doesn't work as well as it should.

this should work but we don;t go in this module when building with jdk19 (because of the issue mentioned above)

**Comment by olamy** [comment]
but definitely the module [jetty-quic-quiche-foreign-incubator](https://github.com/eclipse/jetty.project/tree/jetty-12.0.x/jetty-core/jetty-quic/jetty-quic-quiche/jetty-quic-quiche-foreign-incubator) cannot be build on JDK 19

```


/home/jenkins/agent/workspace/jetty.project_PR-8707/jetty-core/jetty-quic/jetty-quic-quiche/jetty-quic-quiche-foreign-incubator/src/main/java/module-info.java:16: error: module not found: jdk.incubator.foreign

    requires jdk.incubator.foreign;
```

**Comment by olamy** [comment]
obviously the javadoc is failing with jdk19..

```
[ERROR] [parsing started SimpleFileObject[/Volumes/workspace/dev/sources/open-sources/jetty/jetty.project/jetty-core/jetty-quic/jetty-quic-quiche/jetty-quic-quiche-foreign-incubator/src/main/java/org/eclipse/jetty/quic/quiche/foreign/incubator/quiche_h.java]]
[ERROR] [parsing completed 40ms]
[ERROR] /Volumes/workspace/dev/sources/open-sources/jetty/jetty.project/jetty-core/jetty-quic/jetty-quic-quiche/jetty-quic-quiche-foreign-incubator/src/main/java/module-info.java:16: error: module not found: jdk.incubator.foreign
[ERROR]     requires jdk.incubator.foreign;
```

So I remove `javadoc:javadoc` from the jdk19 build but this mean we cannot have jdk19 as a pre requisite for releasing as javadoc will not pass.

**Comment by olamy** [comment]
other issues with jdk19:
- maven plugin its with groovy script 
```
Caused by: java.lang.IllegalArgumentException: Unsupported class file major version 63
	at groovyjarjarasm.asm.ClassReader.<init>(ClassReader.java:199)
```
- infinispan it tests
- http2 tests 
```
Running org.eclipse.jetty.http2.tests.ConnectTimeoutTest.testConnectTimeout()
2022-10-25 01:42:00.373:INFO :oejs.Server:main: jetty-12.0.0-SNAPSHOT; built: 2022-10-25T01:29:41.538Z; git: 79c69aded3b8b1fe1fc0cb50270f1771615e5779; jvm 19+36-2238
2022-10-25 01:42:00.376:INFO :oejs.AbstractConnector:main: Started ServerConnector@60bdf15d{h2, (h2)}{0.0.0.0:38601}
2022-10-25 01:42:00.376:INFO :oejs.Server:main: Started oejs.Server@51e4ccb3{STARTING}[12.0.0-SNAPSHOT,sto=0] @36113ms
2022-10-25 01:42:00.387:WARN :oeji.ManagedSelector:client-407: Could not process key for channel java.nio.channels.SocketChannel[closed]
java.lang.AssertionError: |Expected: an instance of java.net.SocketTimeoutException|     but: <java.net.NoRouteToHostException: No route to host> is a java.net.NoRouteToHostException
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:6)
	at org.eclipse.jetty.http2.tests.ConnectTimeoutTest$3.failed(ConnectTimeoutTest.java:52)
```

**Comment by lorban** [comment]
@olamy maybe we should start using [maven toolchains](https://maven.apache.org/guides/mini/guide-using-toolchains.html) for this module?

FYI, in the near future, we're going to add a comparable module specific to JDK 18 and another one specific to JDK 19.

**Comment by olamy** [comment]
> @olamy maybe we should start using [maven toolchains](https://maven.apache.org/guides/mini/guide-using-toolchains.html) for this module?
> 
> FYI, in the near future, we're going to add a comparable module specific to JDK 18 and another one specific to JDK 19.

OMG what a pain.... (as we control our build image why not but doh...)
do you mean specific to only 18 (e.g `[18]` or `[18,)`? only 18 really? is 18 will be considered as a LTS?

**Comment by lorban** [comment]
We definitely want the JDK 17 module, almost certainly the JDK 19 one, maybe we could give up the JDK 18 one.
@sbordet certainly will want to chime in.

**Comment by joakime** [comment]
> @olamy maybe we should start using [maven toolchains](https://maven.apache.org/guides/mini/guide-using-toolchains.html) for this module?
> 
> FYI, in the near future, we're going to add a comparable module specific to JDK 18 and another one specific to JDK 19.

This sounds like a candidate for our first Muilti-Release JAR file too.

**Comment by olamy** [comment]
closed as already done via other PRs

**Pull request closed without merging**